Asset locking

andriyDev opened the pull request:
# Objective

- Implements #15346.

## Solution

- This is pretty much the same as the design proposed in #15346.

## Testing

- A basic test in `bevy_asset`.

---

## Showcase

```rust
fn my_task_spawning_system(my_mesh_handle: Res<MyMeshHandle>, mut meshes: ResMut<Assets<Mesh>>) {
    let handle: Handle<Mesh> = my_mesh_handle.0.clone();
    let mesh_arc: Arc<Mesh> = meshes.lock(&handle);

    AsyncComputeTaskPool::get().spawn(move async || {
        // Do something with the mesh_arc.
        std::thread::sleep(std::time::Duration::from_secs(mesh_arc.count_vertices()));
        println!("Done processing!");
    });
}
```

The primary new parts of the API are:
- `Assets::lock` - Locks the asset if it exists
- `Assets::try_unlock` - Attempts to unlock the asset if there are no more outstanding handles (otherwise the asset remains locked).

## Migration Guide

- `Assets::get_or_insert_with` may now return an error if the asset is locked.
- `Assets::get_mut` may now return an error if the asset is locked (in addition to if the asset is missing). Consider using `Assets::get_mut_or_clone` if your type is `Clone` to avoid handling locking.
- `Assets::remove` and `Assets::remove_untracked` now return an `AssetPresence<A>`. This requires you to handle the case where an asset is locked.
- `Assets::iter_mut` now returns an `(AssetId<A>, AssetPresenceMut<'a, A>)` instead of `(AssetId<A>, &'a mut A)`. This allows you to handle the locked case.
- `ReflectAsset::get_mut` and `ReflectAsset::get_unchecked_mut` may now return an error if the asset is locked (in addition to if the asset is missing).
- `ReflectAsset::remove` now returns a `ReflectedAssetPresence` to store both the locked and unlocked assets.

andriyDev committed:
Update ReflectAsset to match the AssetPresence API.

andriyDev committed:
Create `AssetPresence` as a replacement to Option.

This currently buys us nothing, but it's a nice stepping stone to making `AssetPresence` have a locked variant.

andriyDev committed:
Add a `Locked` variant to `AssetPresence`.

This variant is still unused.

andriyDev committed:
Allow locking and unlocking assets.

Now assets can be "locked" into an `Arc`.

andriyDev committed:
Flatten out the logic for `DenseAssetStorage::insert`.

andriyDev committed:
Handle locked assets when moving assets from the render world to the main world.

For simplicity, we just clone the assets. In theory, we could have an enum to store the RenderAssets as either a locked Arc or an Unlocked asset, but I'd rather just avoid that branch.

andriyDev committed:
Handle materials being locked when setting wireframe colors.

andriyDev committed:
Add some convenience functions to AssetPresenceMut.

This makes it easier to just unwrap a mutable borrow by just doing `.as_mut().unwrap()`.

andriyDev committed:
Keep track of the set of currently locked assets.

This will make it more performant to try unlocking all the locked assets in the "happy case" that few assets are locked.

andriyDev committed:
Create a system to automatically unlock assets every frame.

This way, if you lock an asset, you just have to drop its `Arc` and wait a frame for the asset to become mutable again (as opposed to requiring a system to figure out which assets can be unlocked and trying to unlock just those).

The price is constant polling whenever an asset is locked. We have to repeatedly try unlocking an asset until it is actually ready to be unlocked.

andriyDev committed:
Update all the examples to handle asset locking.

andriyDev committed:
Handle meshlets possibly being locked.

This is actually a totally fine case, we just need a borrow to be able to clone the internal Arc's.

andriyDev committed:
Add a test of asset locking and unlocking.

andriyDev commented:
Just rebased onto main.

BenjaminBrienen reviewed (approved):
Very clean code and easy to follow!

alice-i-cecile reviewed (changes requested):
The code is really well-made: commented, clearly engineered, great use of enums. Can you add an example demonstrating how (and especially why) a user might use this API?

Once that's in place, I'm on board and am willing to merge.

andriyDev committed:
Create an example for the use of asset locking.

andriyDev commented:
I rebased (#15281 had a big blast radius), and also included a new example to show off how you can use asset locking.

I still need to figure out how to implement `get_mut_or_clone` as mentioned in #15346. I'm running into borrow checker issues (something something polonius something). I'll take another stab at that tomorrow.

andriyDev committed:
Implement `get_mut_or_clone`.

This makes working with asset locking much more palatable, as users don't need to explicitly handle the locking case in most situations.

andriyDev committed:
Use `get_mut_or_clone` in all the examples.

This is likely the common case for users, so we should have our examples use this function.

andriyDev commented:
Ok, I've implemented `get_mut_or_clone`, and all the examples now use it. I've updated the migration guide to point users to `get_mut_or_clone` as well.

One thing that's kinda rough here is that `iter_mut` doesn't allow you to "clone to unlock". I'm not sure if this is worthwhile to solve, it may require a whole new struct to wrap around `AssetPresenceMut`, and that seems a little overcomplicated.

bushrat011899 reviewed (approved):
I think this is a very interesting concept to add to `Assets` and opens up further discussion around other systems. I've left some notes with possible future work, but I think as-is this is good to merge.

andriyDev commented on crates/bevy_asset/src/assets.rs:
```diff
@@ -107,7 +108,155 @@ enum Entry<A: Asset> {
+
+    /// Locks the current asset, which prevents the asset from being modified
+    /// until it is unlocked. Returns [`None`] only if [`AssetPresence::None`].
+    pub fn lock(&mut self) -> Option<Arc<A>> {
```
This is an interesting idea, but it's not clear to me how big a win this is. For one, the list of lock locations can only grow (until an asset is finally unlocked). I suppose if we assume locking is infrequent and short-lived this isn't a big deal, but I can totally see an asset being locked for a long time and constantly getting more and more lock locations. For another, that only tracks the initial lock location, but you're free to clone the Arc as much as you want. I suspect the more dangerous situation is something storing the Arc without knowing it's a locked asset and blocking mutations that way (and there's no real way to solve this AFAICT).

Regardless, this definitely seems worthy of a discussion and I could be convinced!

bushrat011899 replied:
Yeah it's definitely not perfect, but it _could_ help with at least giving the user a starting point for debugging how an asset deadlock started. Potentially, we could even wrap the returned `Arc` to give it better debugging functionality. Storing a `HashMap` of `Location`'s on the heap and counting down each location as the wrapped `Arc` is dropped, and counting up as the `Arc` is cloned.

Definitely half-baked and up for debate, and _absolutely_ out of scope for this PR!

alice-i-cecile commented:
Going to wait on @cart to take another look at this design: he had opinions in the linked thread and I don't feel comfortable merging this without him :)

BenjaminBrienen commented:
Has merge conflicts. Still awaiting response from @cart.
